src/core/exec-invoke.c

```diff
@@ -2563,12 +2601,37 @@ static int setup_exec_directory(
                 if (params->runtime_scope != RUNTIME_SCOPE_SYSTEM)
                         continue;
 
-                /* Then, change the ownership of the whole tree, if necessary. When dynamic users are used we
-                 * drop the suid/sgid bits, since we really don't want SUID/SGID files for dynamic UID/GID
-                 * assignments to exist. */
-                r = path_chown_recursive(pp ?: p, uid, gid, context->dynamic_user ? 01777 : 07777, AT_SYMLINK_FOLLOW);
+                /* Use 'nobody' uid/gid for exec directories if ID-mapping is supported. For backward compatibility,
+                 * continue doing chmod/chown if the directory was chmod/chowned before (if uid/gid is not 'nobody') */
+                bool idmapping_supported = is_idmapping_supported(pp ?: p);
+                log_debug("ID-mapping is%ssupported for exec directory %s", idmapping_supported ? " " : " not ", pp ?: p);
+
+                struct stat st;
+                r = RET_NERRNO(stat(pp ?: p, &st));
                 if (r < 0)
                         goto fail;
+
+                /* Change the ownership of the whole tree, if necessary. When dynamic users are used we
+                * drop the suid/sgid bits, since we really don't want SUID/SGID files for dynamic UID/GID
+                * assignments to exist. */
+                if (uid == 0 || gid == 0) {
+                        i->idmapped = false;
+                        r = path_chown_recursive(pp ?: p, uid, gid, context->dynamic_user ? 01777 : 07777, AT_SYMLINK_FOLLOW);
+                        if (r < 0)
+                                goto fail;
+                } else if (idmapping_supported && st.st_uid == UID_NOBODY && st.st_gid == GID_NOBODY)
+                        i->idmapped = true;
+                else if (exec_directory_is_private(context, type) && idmapping_supported && st.st_uid == (uid_t)0 && st.st_gid == (gid_t)0) {
+                        r = path_chown_recursive(pp ?: p, UID_NOBODY, GID_NOBODY, context->dynamic_user ? 01777 : 07777, AT_SYMLINK_FOLLOW);
+                        if (r < 0)
+                                goto fail;
+                        i->idmapped = true;
+                } else {
+                        i->idmapped = false;
+                        r = path_chown_recursive(pp ?: p, uid, gid, context->dynamic_user ? 01777 : 07777, AT_SYMLINK_FOLLOW);
+                        if (r < 0)
+                                goto fail;
```

Comment:
the first and the last branch of the big series of ifs are identical, which suggests some tweaks of the conditions should be done, to find a way to merge them